src/storage/src/memory.rs

```diff
@@ -33,6 +32,161 @@ use crate::{
+
+#[allow(clippy::type_complexity)]
+pub trait RangeKv: Clone + Send + Sync + 'static {
+    fn range(
+        &self,
+        range: BytesFullKeyRange,
+        limit: Option<usize>,
+    ) -> StorageResult<Vec<(FullKey<Vec<u8>>, Option<Bytes>)>>;
+
+    fn ingest_batch(
+        &self,
+        kv_pairs: impl Iterator<Item = (BytesFullKey, Option<Bytes>)>,
+    ) -> StorageResult<()>;
+
+    fn flush(&self) -> StorageResult<()>;
+}
+
+pub type BTreeMapRangeKv = Arc<RwLock<BTreeMap<BytesFullKey, Option<Bytes>>>>;
+
+impl RangeKv for BTreeMapRangeKv {
+    fn range(
+        &self,
+        range: BytesFullKeyRange,
+        limit: Option<usize>,
+    ) -> StorageResult<Vec<(FullKey<Vec<u8>>, Option<Bytes>)>> {
+        let limit = limit.unwrap_or(usize::MAX);
+        Ok(self
+            .read()
+            .range(range)
+            .take(limit)
+            .map(|(key, value)| (key.to_ref().to_vec(), value.clone()))
+            .collect())
+    }
+
+    fn ingest_batch(
+        &self,
+        kv_pairs: impl Iterator<Item = (BytesFullKey, Option<Bytes>)>,
+    ) -> StorageResult<()> {
+        let mut inner = self.write();
+        for (key, value) in kv_pairs {
+            inner.insert(key, value);
+        }
+        Ok(())
+    }
+
+    fn flush(&self) -> StorageResult<()> {
+        Ok(())
+    }
+}
+
+pub mod sled {
+    use std::fs::create_dir_all;
+
+    use bytes::Bytes;
+    use risingwave_hummock_sdk::key::FullKey;
+
+    use crate::error::StorageResult;
+    use crate::memory::{BytesFullKey, BytesFullKeyRange, RangeKv, RangeKvStateStore};
+
+    #[derive(Clone)]
+    pub struct SledRangeKv {
+        inner: sled::Db,
+    }
+
+    impl SledRangeKv {
+        pub fn new(path: impl AsRef<std::path::Path>) -> Self {
+            SledRangeKv {
+                inner: sled::open(path).expect("open"),
+            }
+        }
+
+        pub fn new_temp() -> Self {
+            create_dir_all("./.risingwave/sled").expect("should create");
+            let path = tempfile::TempDir::new_in("./.risingwave/sled")
+                .expect("find temp dir")
+                .into_path();
+            Self::new(path)
+        }
+    }
+
+    const EMPTY: u8 = 1;
+    const NON_EMPTY: u8 = 0;
+
+    impl RangeKv for SledRangeKv {
+        fn range(
+            &self,
+            range: BytesFullKeyRange,
+            limit: Option<usize>,
+        ) -> StorageResult<Vec<(FullKey<Vec<u8>>, Option<Bytes>)>> {
+            let (left, right) = range;
+            let left = left.map(|key| key.to_ref().encode_reverse_epoch());
+            let right = right.map(|key| key.to_ref().encode_reverse_epoch());
+            let limit = limit.unwrap_or(usize::MAX);
+            let mut ret = vec![];
+            for result in self.inner.range((left, right)).take(limit) {
+                let (key, value) = result?;
+                let full_key = FullKey::decode_reverse_epoch(key.as_ref()).to_vec();
```

Comment:
`sled` seems to not realize the epoch and compare the full key lexicographically. Will this be problematic?

Reply:
Yes, in some corner case where the remaining part of the key after table id is of variable length, it will cause some problem. But seems that there is no good solution for it.

Reply:
I'm afraid there will be some false positives of value mismatching if we use it for `Verify`. 😨

Reply:
@BugenZhao @hzxa21   I have added new logic to filter out the false positive keys after we get and decode the raw keys from sled. A unit test has been added to cover the case. Without this filter, the unit test will fail.